From 89f4be595e0d70b9be77361f9a885b1e9dbaf391 Mon Sep 17 00:00:00 2001
From: Sandeep Malik <Sandeep.Malik@freescale.com>
Date: Tue, 2 Feb 2010 16:46:24 +0530
Subject: [PATCH] talitos: Adds the support for descriptor recycling.

This patch updates the current descriptor allocation and deallocation
methodolgy with a recyling mechanism. This patch allows the descriptors
in talitos driver can be used and avoid un-necessary allocation and
free of descriptors.

Signed-off-by: Sandeep Malik <Sandeep.Malik@freescale.com>
---
 drivers/crypto/talitos.c |   62 ++++++++++++++++++++++++++++++++++++++++------
 1 files changed, 54 insertions(+), 8 deletions(-)

diff --git a/drivers/crypto/talitos.c b/drivers/crypto/talitos.c
index 864e999..86c2fca 100644
--- a/drivers/crypto/talitos.c
+++ b/drivers/crypto/talitos.c
@@ -64,7 +64,8 @@
 
 #define MAP_ARRAY(chan_no)	(3 << (chan_no * 2))
 #define MAP_ARRAY_DONE(chan_no)	(1 << (chan_no * 2))
-
+#define MAX_IPSEC_RECYCLE_DESC 64
+#define MAX_DESC_LEN   160
 
 /* descriptor pointer entry */
 struct talitos_ptr {
@@ -146,6 +147,12 @@ struct talitos_private {
 	u8 core_num_chan[MAX_GROUPS];
 	/* channels numbers of channels mapped to a core */
 	u8 core_chan_no[MAX_GROUPS][MAX_CHAN] ____cacheline_aligned;
+	/* pointer to the cache pool */
+	struct kmem_cache *netcrypto_cache;
+	/* pointer to edescriptor recycle queue */
+	struct talitos_edesc *edesc_rec_queue[NR_CPUS][MAX_IPSEC_RECYCLE_DESC];
+	/* index in edesc recycle queue */
+	u8 curr_edesc[MAX_GROUPS];
 
 	/* request callback napi */
 	struct napi_struct *done_task;
@@ -188,6 +195,34 @@ struct talitos_edesc {
 #define TALITOS_FTR_SRC_LINK_TBL_LEN_INCLUDES_EXTENT 0x00000001
 #define TALITOS_FTR_HW_AUTH_CHECK 0x00000002
 
+struct talitos_edesc *crypto_edesc_alloc(int len, int flags,
+					struct talitos_private *priv)
+{
+	u32 smp_processor_id = smp_processor_id();
+	u32 current_edesc = priv->curr_edesc[smp_processor_id];
+	if (unlikely(current_edesc == 0)) {
+		return kmem_cache_alloc(priv->netcrypto_cache,
+					GFP_KERNEL | flags);
+	} else {
+		 priv->curr_edesc[smp_processor_id] = current_edesc - 1;
+		return priv->edesc_rec_queue[smp_processor_id]
+					[current_edesc - 1];
+	}
+}
+void crypto_edesc_free(struct talitos_edesc *edesc,
+			struct talitos_private *priv)
+{
+	u32 smp_processor_id = smp_processor_id();
+	u32 current_edesc = priv->curr_edesc[smp_processor_id];
+	if (unlikely(current_edesc == (MAX_IPSEC_RECYCLE_DESC - 1))) {
+		kmem_cache_free(priv->netcrypto_cache, edesc);
+	} else {
+		priv->edesc_rec_queue[smp_processor_id][current_edesc] =
+								edesc;
+		priv->curr_edesc[smp_processor_id] = current_edesc + 1;
+	}
+}
+
 static inline unsigned int get_chan_remap(struct talitos_private *priv)
 {
 	return priv->chan_remap;
@@ -1331,6 +1366,7 @@ static void ipsec_esp_encrypt_done(struct device *dev,
 	struct talitos_edesc *edesc;
 	struct scatterlist *sg;
 	void *icvdata;
+	struct talitos_private *priv = dev_get_drvdata(dev);
 
 	edesc = container_of(desc, struct talitos_edesc, desc);
 
@@ -1345,7 +1381,7 @@ static void ipsec_esp_encrypt_done(struct device *dev,
 		       icvdata, ctx->authsize);
 	}
 
-	kfree(edesc);
+	crypto_edesc_free(edesc, priv);
 
 	aead_request_complete(areq, err);
 }
@@ -1360,6 +1396,7 @@ static void ipsec_esp_decrypt_swauth_done(struct device *dev,
 	struct talitos_edesc *edesc;
 	struct scatterlist *sg;
 	void *icvdata;
+	struct talitos_private *priv = dev_get_drvdata(dev);
 
 	edesc = container_of(desc, struct talitos_edesc, desc);
 
@@ -1378,7 +1415,7 @@ static void ipsec_esp_decrypt_swauth_done(struct device *dev,
 			     ctx->authsize, ctx->authsize) ? -EBADMSG : 0;
 	}
 
-	kfree(edesc);
+	crypto_edesc_free(edesc, priv);
 
 	aead_request_complete(req, err);
 }
@@ -1389,6 +1426,7 @@ static void ipsec_esp_decrypt_hwauth_done(struct device *dev,
 {
 	struct aead_request *req = context;
 	struct talitos_edesc *edesc;
+	struct talitos_private *priv = dev_get_drvdata(dev);
 
 	edesc = container_of(desc, struct talitos_edesc, desc);
 
@@ -1399,7 +1437,7 @@ static void ipsec_esp_decrypt_hwauth_done(struct device *dev,
 		     DESC_HDR_LO_ICCR1_PASS))
 		err = -EBADMSG;
 
-	kfree(edesc);
+	crypto_edesc_free(edesc, priv);
 
 	aead_request_complete(req, err);
 }
@@ -1458,6 +1496,7 @@ static int ipsec_esp(struct talitos_edesc *edesc, struct aead_request *areq,
 	unsigned int ivsize = crypto_aead_ivsize(aead);
 	int sg_count, ret;
 	int sg_link_tbl_len;
+	struct talitos_private *priv = dev_get_drvdata(dev);
 
 	/* hmac key */
 	map_single_talitos_ptr(dev, &desc->ptr[0], ctx->authkeylen, &ctx->key,
@@ -1557,7 +1596,7 @@ static int ipsec_esp(struct talitos_edesc *edesc, struct aead_request *areq,
 	ret = talitos_submit(dev, desc, callback, areq);
 	if (ret != -EINPROGRESS) {
 		ipsec_esp_unmap(dev, edesc, areq);
-		kfree(edesc);
+		crypto_edesc_free(edesc, priv);
 	}
 	return ret;
 }
@@ -1596,6 +1635,7 @@ static struct talitos_edesc *talitos_edesc_alloc(struct device *dev,
 	struct talitos_edesc *edesc;
 	int src_nents, dst_nents, alloc_len, dma_len;
 	int src_chained, dst_chained = 0;
+	struct talitos_private *priv = dev_get_drvdata(dev);
 	gfp_t flags = cryptoflags & CRYPTO_TFM_REQ_MAY_SLEEP ? GFP_KERNEL :
 		      GFP_ATOMIC;
 
@@ -1629,7 +1669,7 @@ static struct talitos_edesc *talitos_edesc_alloc(struct device *dev,
 		alloc_len += icv_stashing ? authsize : 0;
 	}
 
-	edesc = kmalloc(alloc_len, GFP_DMA | flags);
+	edesc = crypto_edesc_alloc(alloc_len, GFP_DMA | flags, priv);
 	if (!edesc) {
 		dev_err(dev, "could not allocate edescriptor\n");
 		return ERR_PTR(-ENOMEM);
@@ -1792,12 +1832,13 @@ static void ablkcipher_done(struct device *dev,
 {
 	struct ablkcipher_request *areq = context;
 	struct talitos_edesc *edesc;
+	struct talitos_private *priv = dev_get_drvdata(dev);
 
 	edesc = container_of(desc, struct talitos_edesc, desc);
 
 	common_nonsnoop_unmap(dev, edesc, areq);
 
-	kfree(edesc);
+	crypto_edesc_free(edesc, priv);
 
 	areq->base.complete(&areq->base, err);
 }
@@ -1816,6 +1857,7 @@ static int common_nonsnoop(struct talitos_edesc *edesc,
 	unsigned int cryptlen = areq->nbytes;
 	unsigned int ivsize;
 	int sg_count, ret;
+	struct talitos_private *priv = dev_get_drvdata(dev);
 
 	/* first DWORD empty */
 	desc->ptr[0].len = 0;
@@ -1898,7 +1940,7 @@ static int common_nonsnoop(struct talitos_edesc *edesc,
 	ret = talitos_submit(dev, desc, callback, areq);
 	if (ret != -EINPROGRESS) {
 		common_nonsnoop_unmap(dev, edesc, areq);
-		kfree(edesc);
+		crypto_edesc_free(edesc, priv);
 	}
 	return ret;
 }
@@ -2251,6 +2293,7 @@ static int talitos_remove(struct of_device *ofdev)
 
 	dev_set_drvdata(dev, NULL);
 
+	kmem_cache_destroy(priv->netcrypto_cache);
 	kfree(priv);
 
 	return 0;
@@ -2511,6 +2554,9 @@ static int talitos_probe(struct of_device *ofdev,
 			}
 		}
 	}
+	priv->netcrypto_cache = kmem_cache_create("netcrypto_cache",
+						MAX_DESC_LEN, 0,
+					SLAB_HWCACHE_ALIGN, NULL);
 
 	return 0;
 
-- 
1.5.6.3

